webhook_to_fedora_messaging/move.py

```diff
@@ -0,0 +1,145 @@
+                print(f"An error occurred: {expt}")
+            finally:
+                await sess.close()
+
+
+async def import_repo_to_database():
+    pairdata, done = read_pair(), set()
+    async for sess in get_session():
+        for item in read_repo():
+            name, desc, lang, team, active = item
+            try:
+                flag = 0
+                if team in pairdata:
+                    flag = 1
+                else:
+                    try:
+                        flag = 2
+                        qery = select(User).filter_by(name=team).options(selectinload("*"))
+                        rslt = await sess.execute(qery)
+                        user = rslt.scalar_one()
+                    except NoResultFound:
+                        print(f"[{datetime.now().strftime(format)}] Service '{team}/{name}' abandoned.")
+                        continue
+
+                # SERVICE IS TO BE CREATED ONLY IF FLAG IS 1 OR 2
+                # ORPHANS ARE TO BE ABANDONED
+                make_service = Service(
+                    name=f"{team}/{name}",
+                    uuid=uuid4().hex[0:8],
+                    type="github",
+                    desc=f"Written in {lang}. {desc}" if lang != "None" else f"{desc}",
+                    disabled=False if active == "t" else True
+                )
+                sess.add(make_service)
+                await sess.flush()
+                await sess.commit()
+
+                if flag == 1:
+                    if team not in done:
+                        for username in pairdata[team]:
+                            qery = select(User).filter_by(name=username).options(selectinload("*"))
+                            rslt = await sess.execute(qery)
+                            user = rslt.scalar_one()
+                            stmt = owners_table.insert().values(
+                                {"service_id": make_service.id, "user_id": user.id}
+                            )
+                            await sess.execute(stmt)
+                            await sess.commit()
+                            done.add(team)
+                elif flag == 2:
+                    qery = select(User).filter_by(name=team).options(selectinload("*"))
+                    rslt = await sess.execute(qery)
+                    user = rslt.scalar_one()
+                    stmt = owners_table.insert().values(
+                        {"service_id": make_service.id, "user_id": user.id}
+                    )
+                    await sess.execute(stmt)
+                    await sess.commit()
+
+                print(f"[{datetime.now().strftime(format)}] Service '{team}/{name}' created.")
+            except Exception as expt:
+                await sess.rollback()
+                print(f"[{datetime.now().strftime(format)}] Service '{team}/{name}' failed.")
+                print(f"An error occurred: {expt}")
+            finally:
+                await sess.close()
```

Comment:
This is happening on each iteration of the for loop, no? We could keep the session open for the entire import I think.

Reply:
Gotcha. Let me know if this works.

```python
async def import_user_to_database():
    async for sess in get_session():
        for item in read_user():
            try:
                if "github_org" not in item:
                    make_user = User(name=item)
                    sess.add(make_user)
                    await sess.flush()
                    await sess.commit()
                    print(f"[{datetime.now().strftime(format)}] User '{item}' created.")
                else:
                    print(f"[{datetime.now().strftime(format)}] User '{item}' abandoned.")
            except Exception as expt:
                await sess.rollback()
                print(f"[{datetime.now().strftime(format)}] User '{item}' failed.")
                print(f"An error occurred: {expt}")
        await sess.close()
```

Reply:
Nope - That would not cut it. 

Reverting the changes.

```
An error occurred: (sqlalchemy.dialects.postgresql.asyncpg.InterfaceError) <class 'asyncpg.exceptions._base.InterfaceError'>: cannot perform operation: another operation is in progress
[SQL: INSERT INTO services (token, name, type, "desc", disabled, sent, uuid, creation_date) VALUES ($1::VARCHAR, $2::VARCHAR, $3::VARCHAR, $4::VARCHAR, $5::BOOLEAN, $6::INTEGER, $7::VARCHAR, $8::TIMESTAMP WITH TIME ZONE) RETURNING services.id]
[parameters: ('269830697b5c48bc8817f284b924e477', 'ceph/ceph-nagios-plugins', 'github', 'Written in Python. Nagios plugins for Ceph', True, 0, '2a25c581', datetime.datetime(2024, 8, 22, 2, 57, 46, 487915, tzinfo=datetime.timezone.utc))]
(Background on this error at: https://sqlalche.me/e/20/rvf5)
```

Reply:
Nah. That thing works.

I just had to add a little bit of ...

```python
async def main():
    await import_user_to_database()
    await import_repo_to_database()


if __name__ == "__main__":
    # Ensure that the W2FM_CONFIG environment variable points towards the correct file
    conf = get_config()
    run(main())
```